llama-index-integrations/tools/llama-index-tools-agentql/llama_index/tools/agentql/base.py

```diff
@@ -0,0 +1,227 @@
+        self.async_browser = async_browser
+
+        self.extract_data_timeout = extract_data_timeout
+        self.extract_elements_timeout = extract_elements_timeout
+
+        self.wait_for_page_load = wait_for_page_load
+        self.wait_for_network_idle = wait_for_network_idle
+
+        self.include_hidden_data = include_hidden_data
+        self.include_hidden_elements = include_hidden_elements
+
+        self.response_mode = response_mode
+        self.stealth_mode = stealth_mode
+
+        self.is_scroll_to_bottom_enabled = is_scroll_to_bottom_enabled
+        self.is_screenshot_enabled = is_screenshot_enabled
+
+    @classmethod
+    def from_async_browser(cls, async_browser: AsyncBrowser) -> "AgentQLToolSpec":
+        """
+        Initialize AgentQLToolSpec from an async browser instance.
+        """
+        return cls(async_browser=async_browser)
+
+    @staticmethod
+    async def create_async_playwright_browser(
+        headless: bool = True, args: Optional[List[str]] = None
+    ) -> AsyncBrowser:
+        """
+        Create an async playwright browser.
+
+        Args:
+            headless: Whether to run the browser in headless mode. Defaults to True.
+            args: arguments to pass to browser.chromium.launch
+
+        Returns:
+            AsyncBrowser: The playwright browser.
+        """
+        from playwright.async_api import async_playwright
+
+        browser = await async_playwright().start()
+        return await browser.chromium.launch(headless=headless, args=args)
+
+    async def extract_web_data(
+        self, url: str, query: Optional[str] = None, prompt: Optional[str] = None
+    ) -> dict:
+        """
+        Extract structure data from a web page given an URL.
+        The data is extracted with either a agentql query or a description of the data to extract.
+
+        Args:
+            url: The URL of the web page to extract data from.
+            query: The AgentQL query used to extract the data. The query must be enclosed with curly braces `{}`. Either this field or prompt field must be provided.
+            prompt: The natural language description of the data you want to extract. Either this field or query field must be provided.
+
+        Returns:
+            dict: The extracted data.
+        """
+        # Check that the URL scheme is valid
+        validate_url_scheme(url)
+
+        # Check if one of 'query' or 'prompt' is provided
+        if not query and not prompt:
+            raise ValueError("Either 'query' or 'prompt' must be provided")
+
+        params = {
+            "wait_for": self.wait_for_page_load,
+            "is_scroll_to_bottom_enabled": self.is_scroll_to_bottom_enabled,
+            "mode": self.response_mode,
+            "is_screenshot_enabled": self.is_screenshot_enabled,
+        }
+        metadata = {
+            "experimental_stealth_mode_enabled": self.stealth_mode,
+        }
+        api_key = os.getenv("AGENTQL_API_KEY")
+        if not api_key:
```

Comment:
i'd move this check higher to pair with query/prompt validation above, or even just do the lookup during init and store in the tool as a private var since we can detect even sooner that all calls are going to fail if this env var isn't set

Reply:
Agree w moving to init